README.md
people-in-docker
================


This repository contains a fully integrated stack to run a web server for each UNIX user (see [Apache UserDir](https://httpd.apache.org/docs/2.4/fr/howto/public_html.html)).


## Project

This project is divided in three parts, corresponding to the three subfolders :

 - `nginx-proxy/` : the front-end proxy
 - `node-proxy/`  : a node app managing users' containers
 - `http-server/` : the web server spawned for each UNIX user within a Docker container


## Installation

See [INSTALL.md](INSTALL.md)


## License

All contents licensed under the [WTFPL](LICENSE)


CHANGELOG.md
## Unversioned changes

 - Use EJS as templating language
 - Use wait-on module to check users' containers health
 - Pull docker image if not exists
 - Error handling

## v0.3.1 - 02 August 2016
### REFACTORING

 - Separate functions between router and controller
 - Add a function to force trigger containers' timeouts

## v0.3.0 - 02 August 2016
### FINALLY A FULLY CONFIGURABLE STACK


node-proxy/app.js
/**
 * Created by badouralix on 22/07/16.
 */
'use strict';


/**
 * Load packages
 **********************************************************************************************************************/
var log     = require('winston');
var morgan  = require('morgan');
var express = require('express');
var fs      = require('fs');
var path    = require('path');
var favicon = require('serve-favicon');
var robots  = require('express-robots');

var config                  = require('./config');
var proxy_router            = require('./routes/proxy');
var stop_users_containers   = require('./local_modules/users-container').trigger_timeouts;


/**
 * Base setup
 **********************************************************************************************************************/

// Setup config
var port = config.app.port;

// Setup log config
log.level = config.log.level;
log.remove(log.transports.Console).add(log.transports.Console, { colorize: true });
log.add(log.transports.File, { filename: '/var/log/node/node-proxy.log' });

// Create express server
var app = express();

// Use morgan to print logs
app.use( morgan('combined', {stream: fs.createWriteStream('/var/log/node/access.log', {flags: 'a'})}) );
if ( app.get('env') === 'development' ) {
	app.use( morgan('dev') );
} else {
	app.use( morgan('combined') );
}

// Setup view engine
app.set('views', path.join(__dirname, 'views'));
app.set('view engine', 'ejs');


/**
 * Configure routes
 **********************************************************************************************************************/
app.use( favicon(path.join(__dirname, 'public', 'favicon.ico')) );
app.use( robots(path.join(__dirname, 'public', 'robots.txt')) );
app.use( express.static(path.join(__dirname, 'public')) );

app.use( proxy_router );

// This is the last non-error-handling middleware use()d. Since nothing else responded, we assume 404.
app.use( function (req, res, next) {
	var err = new Error("Not found");
	err.status = 404;

	next(err);
});


/**
 * Error handling
 **********************************************************************************************************************/

// Development error handler
// Will print stacktrace
if ( app.get('env') === 'development' ) {

	app.use( function(err, req, res, next) {
		res.status(err.status || 500);
		res.render('error', {
			message: err.message,
			error: err
		});
	});

}

// Production error handler
// No stacktraces leaked to user
app.use( function(err, req, res, next) {
	res.status(err.status || 500);
	res.render('error', {
		message: err.message,
		error: {}
	});
});


/**
 * Start the server
 **********************************************************************************************************************/
var server = app.listen( port, function () {

	log.info( "Magic happens on port " + port );

}).on('close', function () {

	stop_users_containers();
	log.info( "Magic just stopped" );

});


/**
 * Handle sigterm
 **********************************************************************************************************************/
process.on('SIGTERM', function () {

	server.close();

});


node-proxy/config.js
/**
 * Created by badouralix on 25/07/16.
 */
'use strict';

var os      = require('os');
var config  = {};

config.app          = {};
config.app.port     = process.env.PORT      || 8080;

config.log          = {};
config.log.level    = process.env.LOG_LEVEL || 'info';

config.proxy                    = {};
config.proxy.username_regex     = '^[a-z_][a-z0-9_-]*[$]?$';
config.proxy.uid_min            = 1000;
config.proxy.wait_after_refused = 10;
config.proxy.container_timeout  = 600000;   // default timeout is 10min

config.docker                   = {};
config.docker.socket            = process.env.DOCKER_SOCKET || '/var/run/docker.sock';
config.docker.image_name        = process.env.PEOPLE_CONTAINER_IMAGE            || 'nginx:1.11-alpine';
config.docker.log_suffix        = process.env.PEOPLE_CONTAINER_LOG_SUFFIX       || '-log';
config.docker.container_prefix  = process.env.PEOPLE_CONTAINER_PREFIX           || 'people_user-';
config.docker.labels_prefix     = process.env.PEOPLE_CONTAINER_LABELS_PREFIX    || os.hostname();
config.docker.cmd               = process.env.PEOPLE_CMD                        || './startup.sh';
config.docker.wait_before_kill  = 20;       // number of seconds to wait before killing the container

module.exports = config;


node-proxy/routes/proxy.js
/**
 * Created by badouralix on 22/07/16.
 */
'use strict';

// Create new express router
var express = require('express');
var router  = express.Router();

// Setup config
var config  = require('../config');

// Setup log config
var log = require('winston');

// Load controller
const user_route = require('../local_modules/users-container').user_route;


/**
 * Define router
 **********************************************************************************************************************/

router.param('username', function (req, res, next, username) {

	// Check username syntax according to a regex ( if defined in config.proxy.username_regex ).
	// On Linux-based systems, usernames usually match /^[a-z_][a-z0-9_-]*[$]?$/.

	if (( config.proxy.username_regex !== undefined ) && ( ! username.match(config.proxy.username_regex) )) {

		// If username doesn't match the username_regex, the app responds with a 400 error

		log.warn("Invalid username detected for " + username );

		var err = new Error("Invalid username");
		err.status = 400;
		return next(err);

	} else {

		// Else, it keeps going.
		// Username is saved in req.username in case it is modified.

		req.username = username;
		next();

	}
});

// Any valid route must start with "~username"
router.all('/~:username*', user_route);


/**
 * Export module
 **********************************************************************************************************************/

module.exports = router;


node-proxy/routes/default.js
/**
 * Created by badouralix on 02/08/16.
 */
'use strict';

// Create new express router
var express = require('express');
var router  = express.Router();

// Setup config
var config  = require('../config');

// Setup log config
var log = require('winston');

// Get app directory
var path    = require('path');
var app_dir = path.dirname(require.main.filename);


/**
 * Define router
 **********************************************************************************************************************/

router.get('/icons/:icon', function (req, res) {
    var options = {
        root: app_dir + '/html/icons/',
        dotfiles: 'deny',
        headers: {
            'x-timestamp': Date.now(),
            'x-sent': true
        }
    };

    res.sendFile(req.params.icon, options, function (err) {
        if ( err ) {
            res.sendStatus(err.statusCode);
        }
        else {
            log.verbose("Successfully sent " + req.params.icon);
        }
    });
});

// Other routes are forbidden or not defined
router.all('/', function(req, res) {
    res.sendStatus(403)
});

router.all('*', function(req, res) {
    res.sendStatus(404)
});


/**
 * Export module
 **********************************************************************************************************************/

module.exports = router;


node-proxy/local_modules/users-container.js
/**
 * Created by badouralix on 02/08/16.
 *
 * Controller for users' containers.
 */
'use strict';

// Setup config
var config  = require('../config');

// Setup log config
var log = require('winston');

// Create a proxy server with custom application logic
var http_proxy  = require('http-proxy');
var proxy       = http_proxy.createProxyServer();

var passwd_user     = require('passwd-user');
var wait_on         = require('wait-on');
var docker_wrapper  = require('./docker-wrapper');

// Create a global variable containing all timeout ids
var timeout_id = {};


/*
 * Define useful functions
 **********************************************************************************************************************/

var check_user = function (req, res, next) {
	var username = req.username;

	// Check if existing user matches username
	passwd_user(username).then(function (user) {
		if (user === undefined) {

			log.debug("No user " + username + " has been found -> aborting task!");

			var err = new Error("User " + username + " was not found");
			err.status = 404;
			return next(err);

		} else if ( user.uid < config.proxy.uid_min ) {

			log.warn("User " + user.username + " is not supposed to be accessible -> aborting task!");

			var err = new Error("Forbidden");
			err.status = 403;
			return next(err);

		}

		route_user(req, res, next, user);

	});
};

var route_user = function (req, res, next, user) {


	log.debug("User " + user.username + " has been found -> looking for container..." );
	docker_wrapper.setup_container(user, function (err, ip) {
		if ( err ) {
			return next(err);
		}

		var http_target = 'http://' + ip + ':80';
		var tcp_target  = 'tcp:' + ip + ':80';

		log.silly("Target for " + user.username + " is " + http_target);

		// Wait for the targeted container to be ready and listening on port 80
		wait_on({ resources: [tcp_target], interval: config.proxy.wait_after_refused, window: 0}, function (err) {
			if ( err ) {
				return next(err);
			}

			// Forward client request to the targeted container
			proxy.web(req, res, { target: http_target }, function (err) {
				return next(err);
			});
		});

	});

	update_timeout(user);

};

var update_timeout = function (user) {

	var container_name =  docker_wrapper.get_user_container_name(user);

	if ( timeout_id.hasOwnProperty(user.username) ) {
		log.silly("Updating timeout for container " + container_name);
		clearTimeout(timeout_id[user.username]);
	} else {
		log.silly("Setting timeout for container " + docker_wrapper.get_user_container_name(user) + "");
	}

	// Define the function to call when the timer elapses. It stops user's container and delete the timeout object
	// associated.
	var callback = function () {

		docker_wrapper.stop_container(user, function (err) {
			log.error(err);
		});

		delete timeout_id[user.username];

	};

	// Save timeout object to a global json object.
	timeout_id[user.username] = setTimeout( callback, config.proxy.container_timeout );

};

var trigger_timeouts = function () {

	// This function forces users' containers launched to stop. Theses containers should have been stopped after a
	// certain amount of ms ( defined in config.proxy.container_timeout ), with a setTimeout saved in the json object
	// timeout_id. Timeouts are cleared and containers are stopped.

	for ( var username in timeout_id ) {
		clearTimeout(timeout_id[username]);

		passwd_user(username).then(function (user) {
			if (user !== undefined) {
				docker_wrapper.stop_container(user);
			}
		});
	}
};


/*
 * Export module
 **********************************************************************************************************************/

module.exports = {
    user_route: check_user,
    trigger_timeouts: trigger_timeouts
};


node-proxy/local_modules/docker-wrapper.js
/**
 * Created by badouralix on 28/07/16.
 */
'use strict';

// Setup config
var config  = require('../config');

// Setup log config
var log = require('winston');

var Docker = require('dockerode');
var fs     = require('fs');

var socket = config.docker.socket;
var stats  = fs.statSync(socket);

if (!stats.isSocket()) {
    throw new Error('Are you sure the docker is running?');
}
var docker = new Docker({ socketPath: socket });

// Get network of the node container #L113 ( needs to improve... )
var network_name = '';
docker.getContainer(process.env.HOSTNAME).inspect( function (err, data) {
	network_name = data.HostConfig.NetworkMode;
});


var get_user_container_name = function (user) {
	var container_prefix = config.docker.container_prefix;
	//var container_name = container_prefix + user.username;

	return container_prefix + user.username;
};

var setup_container = function (user, callback) {
	var container_name = get_user_container_name(user);
	log.verbose("Targeting container " + container_name);

	var container = docker.getContainer(container_name);
	container.inspect( user, function (err, data) {
		if ( err ) {
			//throw( err );
			// create_container(container_name, user, callback); // needs to pay attention to 404 error response from docker api
		}

		if ( data == null ) {
			// Container does not exist
			setup_image( function (err) {
				if (err) {
					return callback(err);
				}

				create_container(container_name, user, callback);
			});


		} else if ( data.State.Running == false ) {
			// Container is stopped
			start_container(container, container_name, callback);

		} else if ( data.State.Paused == true ) {
			// Container is paused
			unpause_container(container, container_name, callback);

		} else {
			log.debug("Using started container " + container_name);
			get_ip_from_data(data, callback)

		}

	});
};

var start_container = function (container, container_name, callback) {
	log.info("Starting container " + container_name);

	container.start( function () {
		// weird : start callback doesn't not return updated data and therefore container needs inspect call
		get_ip_from_container(container, callback);
	});
};

var unpause_container = function (container, container_name, callback) {
	log.info("Unpausing container " + container_name);

	container.unpause( function () {
		get_ip_from_container(container, callback);
	});
};

var create_container = function (container_name, user, callback) {
	log.info("Creating container " + container_name);

	var image_name = config.docker.image_name;
	var log_suffix = config.docker.log_suffix;

	var create_options = {
		'name': container_name,
		'labels': {},
		'cmd': config.docker.cmd,
		'image': image_name,
		'workingDir': '/mnt',
		'exposedPorts': {},
		'hostConfig': {
			'binds': [
				user.homedir + ':/home/' + user.username + ':ro',
				process.env.PEOPLE_SYNC_DIR + ':/mnt:ro',
				container_name + log_suffix + ':/var/log',
				'/etc/localtime:/etc/localtime:ro'
			],
			'networkMode': network_name //'container:' + process.env.HOSTNAME doesn't work...
		}
	};
	create_options['labels'][config.docker.labels_prefix + '.automatic_run'] = true.toString();
	create_options['labels'][config.docker.labels_prefix + '.username'] = user.username;

	docker.createContainer( create_options, function (err, container) {
		if ( err ) {
			return callback(err);
		}

		start_container(container, container_name, callback);
	});
};

var setup_image = function (callback) {
	var image_name = config.docker.image_name;
	log.silly("Checking image " + image_name);

	var image = docker.getImage(image_name);
	image.inspect( function (err, data) {

		if ( err && err.statusCode == 404 ) {

			// The image image_name hasn't been found and must be pulled
			pull_image(image_name, callback)

		} else {

			// The image image_name has been found or the error can't be handled here
			callback(err);

		}
	});
};

var pull_image = function (image_name, callback) {
	log.info("Pulling image " + image_name);

	docker.pull(image_name, function (err, stream) {
		if ( err ) {
			callback(err);
		}

		docker.modem.followProgress(stream, on_finished);

		function on_finished(err, output) {
			if ( err ) {
				return callback(err);
			}

			log.silly("Image " + image_name + " has been successfully pulled \\o/");
			return callback(null);
		}
	})
};

var get_ip_from_container = function (container, callback) {
	container.inspect( function (err, data) {
		if ( err ) {
			return callback(err);
		}

		get_ip_from_data(data, callback);
	});
};

var get_ip_from_data = function (data, callback) {
	var container_id = data.Id;
	var networks = data.NetworkSettings.Networks;

	if ( Object.keys(networks).length != 1 ) {
		return callback( new Error("An error occured while parsing networks of container " + container_id + " ( probably too many networks )") );
	} else {
		return callback( null, networks[Object.keys(networks)[0]].IPAddress );
	}
};

var stop_container = function (user, callback) {
	var container_name = get_user_container_name(user);
	log.info("Stopping container " + container_name);

	var container = docker.getContainer(container_name);
	container.inspect( user, function (err, data) {
		if ( err ) {
			return callback(err);
		}

		if ( data.State.Running == true ) {
			// Container is stopped
			container.stop( { t: config.docker.wait_before_kill }, function () {} );
		}
	});

};


module.exports = {
	get_user_container_name: get_user_container_name,
	setup_container: setup_container,
	stop_container: stop_container
};
